### PR TITLE
feat: Introducing telephony module

### PR DESCRIPTION
**Pull request by leela**
Introducing a module named Telephony, that holds all the common doctype's needed to integrate any telephony communication app.

**Added**
* New module named Telephony added.
* `Incoming Call Settings` and `Incoming Call Handling Schedule` doctype's added
* Moved call_log doctype from communication into Telephony module
* Added validations for new doctype's
* Phone icon that is part of frappe is moved into ERP

**Screenshot**
incoming_call_settings

> no-docs

Docs will be added once we are done with one of the telephony integrations.

### CODE REVIEW

File: erpnext/telephony/doctype/incoming_call_settings/incoming_call_settings.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+		}
+	}
+	return false;
+}
+
+function validate_call_schedule_overlaps(schedule) {
+	let group_by_day = groupby(schedule, 'day_of_week');
+	for (const [day, day_schedule] of Object.entries(group_by_day)) {
+		if (is_call_schedule_overlapped(day_schedule)) {
+			frappe.throw(`Please fix overlapping time slots for ${day}.`);
```
Comment (surajshetty3416) [inline erpnext/telephony/doctype/incoming_call_settings/incoming_call_settings.js]:
```suggestion
			frappe.throw(__("Please fix overlapping time slots for {0}.", [day]);
```
User facing strings should be translatable.
https://github.com/frappe/erpnext/wiki/Coding-Standards#user-facing-messages-must-be-translated

More information on translations: https://frappeframework.com/docs/user/en/translations

File: erpnext/telephony/doctype/incoming_call_settings/incoming_call_settings.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+function validate_call_schedule_timeslot(schedule) {
+	// Make sure that to time slot is ahead of from time slot.
+	let errors = [];
+
+	for (let row in schedule) {
+		let record = schedule[row];
+		let from_time_in_secs = time_to_seconds(record.from_time);
+		let to_time_in_secs = time_to_seconds(record.to_time);
+		if (from_time_in_secs >= to_time_in_secs) {
+			errors.push(`Call Schedule Row ${row}: To time slot should always be ahead of From time slot.`);
```
Comment (surajshetty3416) [inline erpnext/telephony/doctype/incoming_call_settings/incoming_call_settings.js]:
```suggestion
			errors.push(__("Call Schedule Row {0}: To time slot should always be ahead of From time slot."), [row]);
```

File: erpnext/telephony/doctype/incoming_call_settings/incoming_call_settings.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+		"""	Make sure that to time slot is ahead of from time slot.
+		"""
+		errors = []
+		for record in schedule:
+			from_time = self.time_to_seconds(record.from_time)
+			to_time = self.time_to_seconds(record.to_time)
+			if from_time >= to_time:
+				errors.append(
+					f'Call Schedule Row {record.idx}: To time slot should always be ahead of From time slot.'
+				)
```
Comment (surajshetty3416) [inline erpnext/telephony/doctype/incoming_call_settings/incoming_call_settings.py]:
```suggestion
				errors.append(_("Call Schedule Row {0}: To time slot should always be ahead of From time slot.").format(record.idx))
```
https://github.com/frappe/erpnext/wiki/Coding-Standards#user-facing-messages-must-be-translated

File: erpnext/telephony/doctype/incoming_call_settings/incoming_call_settings.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+		for day in week_days:
+			timeslots = [(record.from_time, record.to_time) for record in schedule if record.day_of_week==day]
+
+			# convert time in timeslot into an integer represents number of seconds
+			timeslots = sorted(map(lambda seq: tuple(map(self.time_to_seconds, seq)), timeslots))
+			if len(timeslots) < 2: continue
+
+			for i in range(1, len(timeslots)):
+				if self.check_timeslots_overlap(timeslots[i-1], timeslots[i]):
+					frappe.throw(f"Please fix overlapping time slots for {day}.")
```
Comment (surajshetty3416) [inline erpnext/telephony/doctype/incoming_call_settings/incoming_call_settings.py]:
```suggestion
					frappe.throw(_("Please fix overlapping time slots for {0}.").format(day))
```

**Review by surajshetty3416 — CHANGES_REQUESTED** [review]

**Review by surajshetty3416 — APPROVED** [review]

**Comment by ameerbaathar** [comment]
how to use the API for this feature ?

**Comment by nashwan762** [comment]
O

**Pull request merged**